src/gridcoin/beacon.cpp

```diff
@@ -1425,6 +1504,43 @@ bool BeaconRegistry::BeaconDB::erase(uint256 hash)
     return Delete(hash);
 }
 
+// Note that this function uses the shared pointer use_count() to determine whether an element in
+// m_historical is referenced by either the m_beacons or m_pending map and if not, erases it, leaving the backing
+// state in leveldb untouched. Note that the use of use_count() in multithreaded environments must be carefully
+// considered because it is only approximate. In this case it is exact. Access to the entire BeaconRegistry class
+// and everything in it is protected by the cs_main lock and is therefore single threaded. This method of passivating
+// is MUCH faster than searching through m_beacons and m_pending for each element, because they are not keyed by hash.
+//
+// Note that this function acts very similarly to the map erase function with an iterator argument, but with a standard
+// pair returned. The first part of the pair is an iterator to the next element and the second is a boolean as to whether
+// the element was passivated. This is designed to be traversed in a for loop just like the normal map erase in
+// C++11 and later.
+std::pair<BeaconRegistry::HistoricalBeaconMap::iterator, bool>
+    BeaconRegistry::BeaconDB::passivate(BeaconRegistry::HistoricalBeaconMap::iterator& iter)
+{
+    // m-historical itself holds one reference, additional references can be held by m_beacons and m_pending.
+    // If there is only one reference then remove the shared_pointer from m_historical, which will implicitly destroy
+    // the shared_pointer object.
+    if (iter != m_historical.end() && iter->second.use_count() == 1)
+    {
+        iter = m_historical.erase(iter);
+        return std::make_pair(iter, true);
+    }
+    else
+    {
+        LogPrint(BCLog::LogFlags::BEACON, "INFO: %s: Passivate called for historical beacon record with hash %s that "
+                                          "has existing reference count %li. This is expected under certain situations, "
+                                          "such as a forced (re)advertisement, where the new pending beacon is allowed "
+                                          "to expire, while the original is still active.",
+                 __func__,
+                 iter->second->m_hash.GetHex(),
+                 iter->second.use_count());
```

Comment:
If, for some reason, this method is passed the iterator for `m_historical.end()`, it will fall into the `else` clause and dereference an invalid iterator. Is the test for `iter != m_historical.end()` misleading in the preceding condition?

Reply:
This is a valid point and I should remove the condition I think. This function is either used individually with a valid iterator passed in that is not .end(), or used in a for loop (in passivate_db) where it also will never be passed .end() into it. So the condition is misleading.